src/driver-sinowealth.c

```diff
@@ -1368,6 +1368,130 @@ sinowealth_update_macro_from_action(struct ratbag_profile *profile, struct ratba
 	macro->event_count = raw_event_count;
 }
 
+struct sinowealth_device_data {
+	const char fw_version[4];
+	const char *device_name;
+	enum sinowealth_led_format led_type;
+	/* Not necessarily the sensor used on the mouse, some mice may have
+	 * different sensers per the same model.
+	 * We may fall back to this value in the future if we ever find mice
+	 * with broken sensor type field.
+	 */
+	enum sinowealth_sensor sensor_type;
+	uint16_t pid;
+	uint16_t vid;
+	unsigned int button_count;
+};
+
+static const struct sinowealth_device_data sinowealth_supported_devices[] = {
+	/* 258a:0027 devices. */
+	{
+		.fw_version = "V102",
+		.device_name = "Glorious Model O (old firmware)",
+		.led_type = LED_RBG,
+		.sensor_type = SINOWEALTH_SENSOR_PMW3360,
+		.vid = 0x258a,
+		.pid = 0x27,
+		.button_count = 6,
+	},
+	{
+		.fw_version = "V161",
+		.device_name = "G-Wolves Hati HT-M Wired",
+		.led_type = LED_NONE,
+		/* Can also be PWM3389. */
+		.sensor_type = SINOWEALTH_SENSOR_PMW3360,
+		.vid = 0x258a,
+		.pid = 0x27,
+		.button_count = 6,
+	},
+	{
+		.fw_version = "3106",
+		.device_name = "DreamMachines DM5 Blink",
+		.led_type = LED_RGB,
+		.sensor_type = SINOWEALTH_SENSOR_PMW3389,
+		.vid = 0x258a,
+		.pid = 0x27,
+		.button_count = 8,
+	},
+	{
+		.fw_version = "3110",
+		.device_name = "Genesis Xenon 770",
+		.led_type = LED_RGB,
+		.sensor_type = SINOWEALTH_SENSOR_PMW3327,
+		.vid = 0x258a,
+		.pid = 0x27,
+		/* Side buttons are detachable.
+		 * There can be either 3 or 9 buttons on the side.
+		 */
+		.button_count = 5 + 9,
+	},
+
+	/* 258a:0033 devices. */
+	{
+		.fw_version = "V102",
+		.device_name = "Glorious Model D",
+		.led_type = LED_RBG,
+		.sensor_type = SINOWEALTH_SENSOR_PMW3360,
+		.vid = 0x258a,
+		.pid = 0x33,
+		.button_count = 6,
+	},
+
+	/* 258a:0036 devices. */
+	{
+		.fw_version = "V103",
+		.device_name = "Glorious Model O/O- (updated firmware)",
+		.led_type = LED_RBG,
+		.sensor_type = SINOWEALTH_SENSOR_PMW3360,
+		.vid = 0x258a,
+		.pid = 0x36,
+		.button_count = 6,
+	},
+
+	/* 258a:003E devices. */
+	{
+		/* Untested. */
+		/* This matches both Classic and Ace versions. */
+		.fw_version = "3105",
+		.device_name = "G-Wolves Hati HT-S Wired",
+		.led_type = LED_NONE,
+		.sensor_type = SINOWEALTH_SENSOR_PMW3360,
+		.vid = 0x258a,
+		.pid = 0x3e,
+		.button_count = 6,
+	},
+
+	/* Dummy device. Make sure it's the last entry. */
+	{
+		.fw_version = "0000",
+		.device_name = "Generic SinoWealth Mouse",
+		.led_type = LED_RBG,
+		.sensor_type = SINOWEALTH_SENSOR_PMW3360,
+		.vid = 0xffff,
+		.pid = 0xffff,
+		.button_count = 6,
+	},
+};
```

Comment:
fwiw, this should really all be in the `.device` files, we do support driver-specific sections (even if the whole thing is a tad awkward).